File: Java/P3/irrgarten/irrgarten/src/irrgarten/Labyrinth.java

package irrgarten;
import java.util.ArrayList;

/**
 * @brief   Funciones del Laberinto                                                                       
 * @author  Isabel Morro Tabares
 */
 

public class Labyrinth {
    
    // Declaración de atributos
    
    private static final char BLOCK_CHAR = 'X';
    private static final char EMPTY_CHAR = '-';
    private static final char MONSTER_CHAR = 'M';
    private static final char COMBAT_CHAR = 'C';
    private static final char EXIT_CHAR = 'E';
    private static final int ROW = 0;
    private static final int COL = 1;
    
    private int nRows;
    private int nCols;
    private int exitRow;
    private int exitCol;

   
    private Monster[][] monsters;
    private char[][] labyrinth;
    private Player[][] players;
 
    
    // Constructor
    
    public Labyrinth (int nRows, int nCols, int exitRow, int exitCol){
        
        this.nCols = nCols;
        this.nRows = nRows;
        this.exitCol = exitCol;
        this.exitRow = exitRow;
        
        monsters = new Monster [nRows][nCols];
        labyrinth = new char [nRows][nCols];
        players = new Player[nRows][nCols];
        
        for (int i = 0; i < nRows ; i++){
            for (int j = 0; j < nCols; j++){
                
                if (i == 0 || i == nRows-1 || j == 0 || j == nCols-1){
                    labyrinth[i][j] = BLOCK_CHAR; 
                }
                else labyrinth[i][j] = EMPTY_CHAR; 
              
                monsters[i][j] = null;
                players[i][j] = null;
            }
        }
        
        labyrinth [this.exitRow][this.exitCol] = EXIT_CHAR;
    }
    
    // Métodos privados
    
    private boolean posOK (int row, int col){
        return row < nRows && col < nCols && row >= 0 && col >= 0;
    }
    
    private boolean emptyPos (int row, int col){
        return labyrinth[row][col] == EMPTY_CHAR;
    }
    
    private boolean monsterPos (int row, int col){
        return labyrinth[row][col] == MONSTER_CHAR;
    }
    
    private boolean exitPos (int row, int col){
        return labyrinth[row][col] == EXIT_CHAR;
    }
    
    private boolean combatPos (int row, int col){
        return labyrinth[row][col] == COMBAT_CHAR;
    }
    
    private boolean canStepOn (int row, int col){
        return posOK(row,col) && (emptyPos(row,col) || monsterPos(row,col) || exitPos(row,col));
    }
    
    private void updateOldPos(int row, int col){
        if (posOK(row,col)){
            if (combatPos(row, col)) labyrinth[row][col] = MONSTER_CHAR;
            else labyrinth[row][col] = EMPTY_CHAR; 
        }
    }
    
    private int[] dir2Pos (int row, int col, Directions direction){
        
        int resultado[] = new int[2];
        switch (direction) {
            case DOWN:
                resultado[0] = row + 1;
                resultado[1] = col;
                break;
            case UP:
                resultado[0] = row - 1;
                resultado[1] = col;
                break;
            case LEFT:
                resultado[0] = row;
                resultado[1] = col - 1;
                break;
            case RIGHT:
                resultado[0] = row;
                resultado[1] = col + 1;
                break;
        }
        
        return resultado;
    }
    
    private int[] randomEmptyPos (){
        
        int fila, columna;
        
        do {
            fila = Dice.randomPos(nRows);
            columna = Dice.randomPos(nCols);
        } while (!emptyPos(fila, columna));
        
        int resultado[] = new int[2];
        resultado[0] = fila;
        resultado[1] = columna;
        
        return resultado;
    }
    

    private Monster putPlayer2D (int oldRow, int oldCol, int row, int col, Player player){
        
        Monster output = null;
        
        if (canStepOn(row,col)){
            
            if(posOK(oldRow, oldCol)){
                Player p = players[oldRow][oldCol];
                if (p == player){
                    updateOldPos(oldRow, oldCol);
                    players[oldRow][oldCol] = null;
                }
            }
            
            boolean monsterPos = monsterPos(row, col);
            
            if (monsterPos){
                labyrinth[row][col] = COMBAT_CHAR;
                output = monsters[row][col];
            }
            else {
                char number = player.getNumber();
                labyrinth[row][col] = number;
            }
            
            players[row][col] = player;
            player.setPos(row, col); 
        }
        
        return output;
    }
    
    // Métodos públicos

    public void spreadPlayers (ArrayList<Player> players){
        for (Player p : players){
            int[] pos = randomEmptyPos();
            putPlayer2D(-1, -1, pos[ROW], pos[COL], p);
        }
    }
    

    public boolean haveAWinner(){
        return !(players[exitRow][exitCol] == null);
    }
    
    
    public void addMonster (int row, int col, Monster monster){
        if (posOK(row,col)){
            if (emptyPos(row,col)){
                labyrinth[row][col] = MONSTER_CHAR;
                monsters[row][col] = monster;
                monster.setPos(row, col);
            }
        }
    }
        
    
    public Monster putPlayer (Directions direction, Player player){
        int oldRow = player.getRow();
        int oldCol = player.getCol();
        int[] newPos = dir2Pos(oldRow, oldCol, direction);
        Monster monster = putPlayer2D(oldRow, oldCol, newPos[ROW], newPos[COL], player);
        return monster;
    }
    
    public void addBlock (Orientation orientation, int startRow, int startCol, int length) {
        
        int incRow, incCol;
        
        if (orientation == Orientation.VERTICAL){
            incRow = 1;
            incCol = 0;
        }
        else {
            incRow = 0;
            incCol = 1;
        }
        
        int row = startRow;
        int col = startCol;
        
        while (posOK(row,col) && emptyPos(row,col) && length > 0){
            labyrinth[row][col] = BLOCK_CHAR;
            length -= 1;
            row += incRow;
            col += incCol;
        }
    }
    
    public ArrayList<Directions> validMoves(int row, int col){
       
        ArrayList<Directions> output = new ArrayList<>();
        
        if (canStepOn(row+1, col)) output.add(Directions.DOWN);
        if (canStepOn(row-1, col)) output.add(Directions.UP);
        if (canStepOn(row, col+1)) output.add(Directions.RIGHT);
        if (canStepOn(row, col-1)) output.add(Directions.LEFT);
        
        return output;
    }
    
    // To String
    @Override
    public String toString(){
        
        String cadena = "";
        for (int i = 0; i < nRows; i++){
            for (int j = 0; j < nCols; j++){
                cadena += labyrinth[i][j];
                if (j == nCols -1) cadena += "\n";
            }
        }
        
        return cadena;
    }
}

File: Ruby/P4/Irrgarten/labyrinth.rb
#encoding:utf-8

=begin

    Autor: Isabel Morro Tabares 

    Funciones del Laberinto
    
=end

require_relative 'dice'
require_relative 'monster'
require_relative 'shield'
require_relative 'orientation'
require_relative 'directions'
require_relative 'fuzzy_player'

module Irrgarten
    class Labyrinth

    private

        # Atributos de clase

        @@BLOCK_CHAR = 'X'
        @@EMPTY_CHAR = '-'
        @@MONSTER_CHAR = 'M'
        @@COMBAT_CHAR = 'C'
        @@EXIT_CHAR = 'E'
        @@ROW = 0
        @@COL = 1

        # Métodos privados

        def pos_ok (row, col)
            row < @n_rows && col < @n_cols && row >= 0 && col >= 0
        end

        def empty_pos (row, col)
            @labyrinth[row][col] == @@EMPTY_CHAR
        end

        def monster_pos(row, col)
            @labyrinth[row][col] == @@MONSTER_CHAR
        end

        def exit_pos (row, col)
            @labyrinth[row][col] == @@EXIT_CHAR
        end

        def combat_pos (row, col)
            @labyrinth[row][col] == @@COMBAT_CHAR
        end

        def can_step_on (row, col)
            pos_ok(row, col) && (empty_pos(row,col) || monster_pos(row,col) || exit_pos(row,col))
        end

        def update_old_pos (row, col)
            if pos_ok(row,col)
                if combat_pos(row, col)
                    @labyrinth[row][col] = @@MONSTER_CHAR
                else 
                    @labyrinth[row][col] = @@EMPTY_CHAR
                end
            end
        end

        def dir_2_pos (row, col, direction)
            
            resultado = Array.new

            case direction
                when Directions::DOWN
                    resultado[0] = row + 1
                    resultado[1] = col
                when Directions::UP
                    resultado[0] = row - 1
                    resultado[1] = col
                when Directions::LEFT
                    resultado[0] = row
                    resultado[1] = col - 1
                else Directions::RIGHT
                    resultado[0] = row
                    resultado[1] = col + 1 
            end
            
            resultado
        end

        def random_empty_pos

            begin
                fila = Dice.random_pos(@n_rows)
                columna = Dice.random_pos(@n_cols)
            end while !(empty_pos(fila, columna))

            resultado = Array.new
            resultado[0] = fila
            resultado[1] = columna

            resultado

        end

        def put_player_2D (old_row, old_col, row, col, player)

            output = nil

            if (can_step_on(row,col))

                if (pos_ok(old_row, old_col))
                    p = @players[old_row][old_col]
                    if (p == player)
                        update_old_pos(old_row, old_col)
                        @players[old_row][old_col] = nil
                    end
                end 

                monster_pos = monster_pos(row, col)

                if (monster_pos)
                    @labyrinth[row][col] = @@COMBAT_CHAR
                    output = @monsters[row][col]
                else 
                    number = player.number
                    @labyrinth[row][col] = number
                end

                @players[row][col] = player

                player.set_pos(row, col)

            end

            output
        end

    public

        # Constructor

        def initialize (n_rows, n_cols, exit_row, exit_col)

            @n_rows = n_rows
            @n_cols = n_cols
            @exit_row = exit_row
            @exit_col = exit_col

            @monsters = Array.new(@n_rows)
            for i in (0..@n_rows-1)
                @monsters[i] = Array.new(@n_cols)
            end
            @players = Array.new(@n_rows)
            for i in (0..@n_rows-1)
                @players[i] = Array.new(@n_cols)
            end
            @labyrinth = Array.new(@n_rows)
            for i in (0..@n_rows-1)
                @labyrinth[i] = Array.new(@n_cols)
            end

            for i in (0..@n_rows-1) 
                for j in  (0..@n_cols-1)
                    
                    if (i == 0 || i == @n_rows-1 || j == 0 || j == @n_cols-1)
                        @labyrinth[i][j] = @@BLOCK_CHAR
                    else
                        @labyrinth[i][j] = @@EMPTY_CHAR
                    end 

                    @monsters[i][j] = nil
                    @players[i][j] = nil

                end 
            end

            @labyrinth[@exit_row][@exit_col] = @@EXIT_CHAR

        end

        # Métodos públicos

        def spread_players (players)

            players.each do |p| 
                pos = random_empty_pos()
                put_player_2D(-1, -1, pos[@@ROW], pos[@@COL], p)
            end

        end

        def have_a_winner
            !(@players[@exit_row][@exit_col] == nil)
        end

        def add_monster (row, col, monster)
            if pos_ok(row,col)
                if empty_pos(row,col)
                    @labyrinth[row][col] = @@MONSTER_CHAR
                    @monsters[row][col] = monster
                    monster.set_pos(row, col)
                end
            end
        end

        def put_player (direction, player)

            old_row = player.row
            old_col = player.col

            new_pos = dir_2_pos(old_row, old_col, direction)
            monster = put_player_2D(old_row, old_col, new_pos[@@ROW], new_pos[@@COL], player)
            return monster

        end

        def add_block (orientation, start_row, start_col, length)

            if (orientation == Orientation::VERTICAL)
                inc_row = 1
                inc_col = 0
            else
                inc_row = 0
                inc_col = 1
            end 

            row = start_row
            col = start_col

            while (pos_ok(row,col) && empty_pos(row,col) && length > 0)
                @labyrinth[row][col] = @@BLOCK_CHAR
                length -=1
                row += inc_row
                col += inc_col
            end

        end


        def valid_moves(row, col)
            
            output = Array.new

            if (can_step_on(row+1, col))
                output << Directions::DOWN 
            end

            if (can_step_on(row-1, col))
                output << Directions::UP 
            end

            if (can_step_on(row, col+1))
                output << Directions::RIGHT 
            end

            if (can_step_on(row, col-1))
                output << Directions::LEFT 
            end

            return output
            
        end

        def resurrection_fuzzy_player (row, col, fp)
            @players[row][col] = fp
        end

        def to_s
            cadena = ""
            for i in (0..@n_rows-1)
                for j in (0..@n_cols-1)
                    cadena << @labyrinth[i][j]
                    if j == @n_cols-1
                        cadena << "\n"
                    end
                end
            end
            cadena
        end

    end
end